notebooks/streaming/streaming_data_preparation.ipynb

```diff
@@ -0,0 +1,212 @@
+    "\n",
+    "### Pre-requisites\n",
+    "\n",
+    "1. If you haven't already, please download LCL dataset from [data.london.gov.uk](https://data.london.gov.uk/dataset/smartmeter-energy-use-data-in-london-households).\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load_ext autoreload\n",
+    "%autoreload 2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from litdata import optimize\n",
+    "import pandas as pd\n",
+    "from pathlib import Path\n",
+    "import torch\n",
+    "from tqdm import tqdm_notebook as tqdm\n",
+    "\n",
+    "from opensynth.data_modules.lcl_data_module import TrainingData\n",
+    "\n",
+    "RANDOM_STATE = 0\n",
+    "OUTPUT_DIR = Path(\"../../data/processed/historical/stream\")\n",
+    "SIZE_LIMIT = \"100mb\"   # size of each chunk pre-compression\n",
+    "COMPRESSION = \"zstd\"   # compression algorithm\n",
+    "NUM_WORKERS = 4        # number of workers for parallel processing"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 💿 Loading LCL Data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data_path = Path(\"../../data/processed/historical/train/lcl_data.csv\")\n",
+    "stats_path = Path(\"../../data/processed/historical/train/mean_std.csv\")\n",
+    "outlier_path = Path(\"../../data/processed/historical/train/outliers.csv\")\n",
+    "\n",
+    "data = pd.read_csv(data_path)\n",
+    "outliers = pd.read_csv(outlier_path)\n",
+    "\n",
+    "# Combine and shuffle data\n",
+    "df = pd.concat([data, outliers])\n",
+    "df = df.sample(\n",
+    "    frac=1, random_state=RANDOM_STATE\n",
+    ").reset_index(drop=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Work out some stats to partition the source data\n",
+    "# Note: this is not strictly necessary for the LCL data as it is so small\n",
+    "# but I'm showing the process here as it is necessary for larger datasets\n",
+    "\n",
+    "NUM_GROUPS = 5\n",
+    "GROUP_SIZE = len(df) // NUM_GROUPS"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This is the function that will be called by litdata.optimize to yield the data in the correct format\n",
+    "# Some things to take into account:\n",
+    "# - Optimize supports parallel processing, so we pass in a range of values that allow us to slice the data\n",
+    "# - Our streaming dataloaders expect the data to be in a particular format (TrainingData)\n",
+    "\n",
+    "def yield_data(idx: int):\n",
+    "    data_slice = df.iloc[idx * GROUP_SIZE: (idx + 1) * GROUP_SIZE]\n",
```

Comment:
Just one note here that you have:

```python
def yield_data(idx: int):
    data_slice = df.iloc[idx * GROUP_SIZE: (idx + 1) * GROUP_SIZE]

    for row in data_slice.itertuples():
        features: dict[str, torch.Tensor] = {
            "month": getattr(row, "month"),
            "dayofweek": getattr(row, "dayofweek"),
        }
        yield TrainingData(
            kwh=getattr(row, "kwh"),
            features=features,
        )
```
this `df` object is in the global scope - IDK if that is the intention!

Reply:
yeah it's because in Faraday I actually fetch the data as well in that function, rather than externally...

`optimize` needs a function with one arg (`idx`), so I could use `partial` to pass in the df as well - might be overkill though?

Reply:
I think using partial is probably better/ makes the code clearer/ less prone to bug, especially if end users might be using the code interactively in a notebook and it's accidentally grabbing the wrong global variable.

Reply:
I did it!